From 7a7e9f91081947ddac1317e40373aa8b3d7e1d1a Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Thu, 10 Oct 2024 10:43:01 +0200
Subject: [PATCH 1282/1695] plat-stm32mp1: add stm32mp1_pa_or_sram_alias_pa()

Add stm32mp1_pa_or_sram_alias_pa() helper function to ease handling
SRAMx physical addresses that could be aliased

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 43167130ca5d7d9ecc802aca1e436f3510d59fb4 plat-stm32mp1: add stm32mp1_pa_or_sram_alias_pa()]
Change-Id: Ied0b94e02274c18c1944a6587c0c8003a490f23d
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/410582
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 core/arch/arm/plat-stm32mp1/main.c            | 22 +++++++++++++++++++
 core/arch/arm/plat-stm32mp1/platform_config.h |  5 +++++
 core/arch/arm/plat-stm32mp1/stm32_util.h      | 12 ++++++++++
 3 files changed, 39 insertions(+)

diff --git a/core/arch/arm/plat-stm32mp1/main.c b/core/arch/arm/plat-stm32mp1/main.c
index 9268b7f9c..ead6aa16c 100644
--- a/core/arch/arm/plat-stm32mp1/main.c
+++ b/core/arch/arm/plat-stm32mp1/main.c
@@ -856,3 +856,25 @@ bool stm32mp_allow_probe_shared_device(const void *fdt, int node)
 
 	return false;
 }
+
+#if defined(CFG_STM32MP15) && defined(CFG_WITH_PAGER)
+paddr_t stm32mp1_pa_or_sram_alias_pa(paddr_t pa)
+{
+	/*
+	 * OP-TEE uses the alias physical addresses of SRAM1/2/3/4,
+	 * not the standard physical addresses. This choice was initially
+	 * driven by pager that needs physically contiguous memories
+	 * for internal secure memories.
+	 */
+	if (core_is_buffer_inside(pa, 1, SRAM1_ALT_BASE, SRAM1_SIZE))
+		pa += SRAM1_BASE - SRAM1_ALT_BASE;
+	else if (core_is_buffer_inside(pa, 1, SRAM2_ALT_BASE, SRAM2_SIZE))
+		pa += SRAM2_BASE - SRAM2_ALT_BASE;
+	else if (core_is_buffer_inside(pa, 1, SRAM3_ALT_BASE, SRAM3_SIZE))
+		pa += SRAM3_BASE - SRAM3_ALT_BASE;
+	else if (core_is_buffer_inside(pa, 1, SRAM4_ALT_BASE, SRAM4_SIZE))
+		pa += SRAM4_BASE - SRAM4_ALT_BASE;
+
+	return pa;
+}
+#endif
diff --git a/core/arch/arm/plat-stm32mp1/platform_config.h b/core/arch/arm/plat-stm32mp1/platform_config.h
index 483d32a29..ef109a2d1 100644
--- a/core/arch/arm/plat-stm32mp1/platform_config.h
+++ b/core/arch/arm/plat-stm32mp1/platform_config.h
@@ -320,6 +320,11 @@
 #define SRAM4_BASE			U(0x30050000)
 #define SRAM4_SIZE			U(0x10000)
 
+#define SRAM1_ALT_BASE			U(0x10000000)
+#define SRAM2_ALT_BASE			U(0x10020000)
+#define SRAM3_ALT_BASE			U(0x10040000)
+#define SRAM4_ALT_BASE			U(0x10050000)
+
 /* RETRAM layout */
 #define RETRAM_BASE			U(0x38000000)
 #define RETRAM_SIZE			U(0x10000)
diff --git a/core/arch/arm/plat-stm32mp1/stm32_util.h b/core/arch/arm/plat-stm32mp1/stm32_util.h
index 6cedcf4cd..77569d02a 100644
--- a/core/arch/arm/plat-stm32mp1/stm32_util.h
+++ b/core/arch/arm/plat-stm32mp1/stm32_util.h
@@ -346,4 +346,16 @@ TEE_Result stm32_activate_internal_tamper(int id);
 
 bool stm32mp_allow_probe_shared_device(const void *fdt, int node);
 
+#if defined(CFG_STM32MP15) && defined(CFG_WITH_PAGER)
+/*
+ * Return the SRAM alias physical address related to @pa when applicable or
+ * @pa if it does not relate to an SRAMx non-aliased memory address.
+ */
+paddr_t stm32mp1_pa_or_sram_alias_pa(paddr_t pa);
+#else
+static inline paddr_t stm32mp1_pa_or_sram_alias_pa(paddr_t pa)
+{
+	return pa;
+}
+#endif
 #endif /*__STM32_UTIL_H__*/
-- 
2.47.2

